pkg/intoto/metadata.go

```diff
@@ -9,89 +9,148 @@ import (
 	canonicaljson "github.com/docker/go/canonical/json"
 )
 
-// Metadata represents the In-Toto metadata stored in TUF.
-// All fields are represented as []byte in order to be stored in the Custom field for TUF metadata.
+// Metadata points to root layout, its public keys, and/or links
 type Metadata struct {
-	// TODO: remove this once the TUF targets key is used to sign the root layout
-	Key    []byte            `json:"key"`
-	Layout []byte            `json:"layout"`
-	Links  map[string][]byte `json:"links"`
+	Data       []byte   `json:"data"`
+	PublicKeys []string `json:"pubkeys"` // filenames
+	Links      []string `json:"links"`   // filenames
 }
 
-// WriteMetadataFiles writes the content of a metadata object into files in a directory
-func WriteMetadataFiles(m *Metadata, dir string) error {
-	abs, err := filepath.Abs(dir)
-	if err != nil {
-		return err
-	}
+// Custom is a generic structure that contains in-toto Metadata
+type Custom struct {
+	InToto Metadata `json:"intoto"`
+}
 
-	//FIXME: no need to actually write filename.
-	err = ioutil.WriteFile(filepath.Join(abs, "root.layout"), m.Layout, ReadOnlyMask)
-	if err != nil {
-		return err
-	}
+// PublicKeys is a map from the GUN-qualified filename
+// (e.g., "example.com/example-org/example-bundle/in-toto-pubkeys/keyid.pub")
+// to a Custom struct
+type PublicKeys map[string]Custom
+
+// RootLayout maps a GUN-qualified filename
+// (e.g., "example.com/example-org/example-bundle/in-toto-metadata/root.layout")
+// to a Custom struct
+type RootLayout struct {
+	Filename string
+	Custom   Custom
+}
+
+// Links is a map from the GUN-qualified filename
+// (e.g., "example.com/example-org/example-bundle/in-toto-metadata/DIGEST/step.link")
+// to a Custom struct
+type Links map[string]Custom
 
-	//FIXME: no need to actually write filenames.
-	err = ioutil.WriteFile(filepath.Join(abs, "root.layout.pub"), m.Key, ReadOnlyMask)
+// GetRawMessage transforms a Custom struct into a raw Canonical JSON message
+func (custom *Custom) GetRawMessage() (canonicaljson.RawMessage, error) {
+	marshalled, err := canonicaljson.MarshalCanonical(custom)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("cannot encode custom metadata into canonical json: %v", err)
```

Comment:
suggestion: this error message will get bubbled up into CNAB tooling eventually, so it would help to say "in-toto custom metadata" so that it's more clear where things went wrong. Otherwise custom metadata means something completely different in cnab (the custom section in the bundle.json).